src/buildstream/_artifactelement.py

```diff
@@ -131,9 +131,9 @@ def clear_artifact_name_cache(cls):
     # Once we've finished pulling an artifact, we assume the
     # state of the pulled artifact.
     #
-    def _pull_done(self):
-        super()._pull_done()
+    def _load_artifact_done(self):
```

Comment:
I think we need to streamline how `ArtifactElement` instances are initialized with this.

This branch changes a bit the semantics of how an artifact is loaded onto an element, but does not appear to update `Element.__initialize_from_artifact()`, perhaps we can have `Element.__initialize_from_artifact()` directly call `Element._load_artifact()` or such.

Another point, is that loading the artifact from the local cache is obviously a hard requirement when loading any `ArtifactElement`, regardless of whether `Stream.query_cache()` was called on the element.

I think this branch does not exactly change the current initialization (and deferring the artifact loading is planned in a later branch), in any case, it should be clear somehow (perhaps in the `Stream.query_cache()` API comments) that this will end up being a `noop` for `ArtifactElements` which are anyway synchronously loaded at instantiation time.

Reply:
Using the `_load_artifact()` code path for artifact element early loading seems sensible. I've pushed a separate commit for this.

In `Stream.query_cache()` I've added an explicit no-op case if the artifact has already been loaded.

Reply:
Thanks :)

My concern was mostly to ensure that the state of an artifact loaded early on, or an artifact loaded as a result of pulling it, would be as close as possible.

Reply:
Looking at your patch, it seems the code has changed more than I expected it might have... anyway, looks good to me :)